wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs

```diff
@@ -0,0 +1,270 @@
+        ask >>= return . ByteArray.convert
+                       . encodeUtf8 @Text @ByteString
+                       . sformat build
+                       . hash
+
+{-------------------------------------------------------------------------------
+  Estimating fees
+-------------------------------------------------------------------------------}
+
+data EstimateFeesError = EstFeesTxCreationFailed NewTransactionError
+
+instance Buildable EstimateFeesError where
+    build (EstFeesTxCreationFailed newTxErr) =
+        bprint ("EstFeesTxCreationFailed " % build) newTxErr
+
+instance Arbitrary EstimateFeesError where
+    arbitrary = EstFeesTxCreationFailed <$> arbitrary
+
+estimateFees :: ActiveWallet
+             -> IO Address
+             -- ^ A computation to generate a new change 'Address.
+             -> (Address -> Either CoinSelHardErr SafeSigner)
+             -- ^ A function to sign each Address
+             -> CoinSelectionOptions
+             -- ^ The options describing how to tune the coin selection.
+             -> HdAccountId
+             -- ^ The source @(root, account) from where the payment was
+             -- originated
+             -> NonEmpty (Address, Coin)
+             -- ^ The payees
+             -> IO (Either EstimateFeesError Coin)
+estimateFees activeWallet@ActiveWallet{..} genChangeAddr signAddress options accountId payees = do
+    snapshot         <- getWalletSnapshot walletPassive
+    let originalUtxo = accountAvailableUtxo snapshot accountId
+    res <- newTransaction activeWallet genChangeAddr signAddress options accountId payees
+    case res of
+         Left e  -> return . Left . EstFeesTxCreationFailed $ e
+         Right tx -> -- calculate the fee as the difference between inputs and outputs.
+             -- NOTE(adn) Apparently we shouldn't worry about the 'ExpenseRegulation'
+             -- affecting the way we sum, as no matter who pays for the fee,
+             -- the final difference won't be affected.
+             return $ Right
+                    $ sumOfInputs tx originalUtxo `unsafeSubCoin` sumOfOutputs tx
+  where
+      sumOfInputs :: TxAux -> Utxo -> Coin
+      sumOfInputs tx utxo =
+          let inputs = Set.fromList $ toList . _txInputs . taTx $ tx
+          in utxoBalance (utxoRestrictToInputs inputs utxo)
+
+      sumOfOutputs :: TxAux -> Coin
+      sumOfOutputs tx =
+          let outputs = _txOutputs . taTx $ tx
+          in paymentAmount outputs
+
+
+-- NOTE(adn) At the moment we are passing
+-- the full set of Hd wallets as input, which means our lookup function
+-- for an 'HdAddress' can span across the whole wallets and can be very
+-- costly. However, in the way the 'DB' is modelled at the moment, we
+-- store the addresses in a \"flat\" representation (i.e. in an 'IxSet'),
+-- so indexing directly by 'Core.Address' might not be any less efficient
+-- than performing two lookups, one on 'HdAccountId' followed by one on
+-- 'Core.Address'.
```

Comment:
I wouldn't worry about efficiency here. However, passing in a substructure of the DB does eliminate potential bugs as  it makes it clearer what we should do.